src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -0,0 +1,291 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
+{
+    using ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>;
+    using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
+
+    /// <summary>
+    /// Struct containing information that exists on edges in a call graph.
+    /// </summary>
+    public struct CallGraphEdge
```

Comment:
I took a look at how the call graph is being used in #395. It looks like there are unused methods (`GetDirectDependencies` and `GetAllDependencies`). The rest of the structs, classes and methods look like they can be internal instead of public, since they are only used by the tests, and possibly might only be used by other transformations in this project (e.g. monomorphization) in the future?

Reply:
Getting dependencies is a required feature of the call graph and will be needed to interact with the call graph. It is too early to say definitively what the access modifiers should be, but I think all of the ones listed as public are supposed to be public, as in we want to make them available to be used by 3rd party developers.

Reply:
I think that, depending on how the graph is used, getting specific dependencies/edges in the graph is not necessarily a requirement. For example, there are higher-level methods that can be used instead, like `GetCallCycles`. Do you have an example of when the public methods need to be used by a third party, or by other projects in the compiler/QDK?

Since there is uncertainty in what the access modifiers should be, I recommend making them internal until there is a need for them to be public.

Reply:
Leaving the access modifiers as public for now.

Reply:
If they stay public, I would recommend adding TODO comments as a reminder to check the access modifiers later and make them more restrictive if possible. (It would be fine with me to have either a TODO comment before each method or one TODO at the top of the file.) Without changing the access modifiers or adding some reminder to the code, I'm not sure the issue is resolved from my point of view.

Reply:
Added a ToDo for the file for access modifiers.